src/cipher.jl

```diff
@@ -33,120 +17,146 @@ immutable NettleCipher
-        ctx = Array(UInt8, ctx_size($name))
-        if nettle_major_version >= 3
-            ccall($fptr_set_encrypt_key, Void, (Ptr{Void}, Ptr{UInt8}),
-                  ctx, pointer(key))
-        else
-            ccall($fptr_set_encrypt_key, Void, (Ptr{Void}, Cuint, Ptr{UInt8}),
-                  ctx, length(key), pointer(key))
+function get_cipher_types()
+    # If we have already gotten the hash types from libnettle, don't query again
+    if isempty(_cipher_types)
+        cipher_idx = 1
+        # nettle_ciphers is an array of pointers ended by a NULL pointer, continue reading hash types until we hit it
+        while( true )
+            ncptr = unsafe_load(cglobal(("nettle_ciphers",nettle),Ptr{Ptr{Void}}),cipher_idx)
+            if ncptr == C_NULL
+                break
+            end
+            cipher_idx += 1
+            nc = unsafe_load(convert(Ptr{NettleCipher}, ncptr))
+            cipher_type = CipherType(nc)
+            _cipher_types[cipher_type.name] = cipher_type
         end
-        CipherEncrypt{$name}(ctx)
     end
+    return _cipher_types
+end
 
-    @eval function CipherDecrypt(::Type{$name},key::Union(AbstractString,Vector{UInt8}))
-        length(key) != key_size($name) && error("Key must be $(key_size($name)) bytes long")
-        ctx = Array(UInt8, ctx_size($name))
-        if nettle_major_version >= 3
-            ccall($fptr_set_decrypt_key, Void, (Ptr{Void}, Ptr{UInt8}),
-                  ctx, pointer(key))
-        else
-            ccall($fptr_set_decrypt_key, Void, (Ptr{Void}, Cuint, Ptr{UInt8}),
-                  ctx, length(key), pointer(key))
-        end
-        CipherDecrypt{$name}(ctx)
-    end
 
-    @eval function decrypt(state::CipherDecrypt{$name},source)
-        result = Array(UInt8,length(source))
-        decrypt!(state,result,source)
-        result
+function Encryptor(name::AbstractString, key)
+    cipher_types = get_cipher_types()
+    name = uppercase(name)
+    if !haskey(cipher_types, name)
+        throw(ArgumentError("Invalid cipher type $name: call Nettle.get_cipher_types() to see available list"))
     end
+    cipher_type = cipher_types[name]
 
-    @eval function decrypt!(state::CipherDecrypt{$name},dst::Vector{UInt8},source::Vector{UInt8})
-        n = length(source)
-        @assert length(dst) == n
-        ccall($fptr_decrypt,Void,(Ptr{Void},Csize_t,Ptr{UInt8},Ptr{UInt8}),state.ctx,sizeof(source),pointer(dst),pointer(source))
-        dst
+    if length(key) != cipher_type.key_size
+        throw(ArgumentError("Key must be $(cipher_type.key_size) bytes long"))
     end
 
-    @eval function encrypt(state::CipherEncrypt{$name},source::Vector{UInt8})
-        result = Array(UInt8,length(source))
-        encrypt!(state,result,source)
-        result
+    state = Array(UInt8, cipher_type.context_size)
+    if nettle_major_version >= 3
+        ccall( cipher_type.set_encrypt_key, Void, (Ptr{Void}, Ptr{UInt8}), state, pointer(key))
```

Comment:
Does this assume that the `key` is NUL-terminated, since I don't see a length being passed?  Maybe use `Cstring` instead of `Ptr{UInt8}` as the argument type then?